internal/ingest/priority_queue.go

```diff
@@ -0,0 +1,95 @@
+package ingest
+
+import "sync"
+
+type Queue struct {
+	lk     sync.Mutex
+	states map[providerID]adInfo
+	order  []providerID
+
+	// used if someone is pulling an empty queue
+	waiters    uint
+	notifyChan chan struct{}
+	doneChan   chan struct{}
+}
+
+func NewPriorityQueue() *Queue {
+	return &Queue{
+		states:     make(map[providerID]adInfo),
+		order:      []providerID{},
+		waiters:    0,
+		notifyChan: make(chan struct{}),
+		doneChan:   make(chan struct{}),
+	}
+}
+
+func (q *Queue) Push(p providerID, a adInfo) {
+	q.lk.Lock()
+	defer q.lk.Unlock()
+
+	if _, ok := q.states[p]; !ok {
+		q.order = append(q.order, p)
+	}
+	q.states[p] = a
+
+	if q.waiters > 0 {
+		q.notifyChan <- struct{}{}
+	}
+}
+
+func (q *Queue) Pop() providerID {
+	q.lk.Lock()
+
+	for len(q.order) == 0 {
+		q.waiters++
+		q.lk.Unlock()
+		select {
+		case <-q.doneChan:
+			return ""
+		case <-q.notifyChan:
+		}
+		q.lk.Lock()
+		q.waiters--
+	}
+
+	p := q.order[0]
+	q.order = q.order[1:]
```

Comment:
Each time `q.order` is advanced by `q.order = q.order[1:]` and then appended to, this will allocate new underlying memory and copy the contents of the old memory into the new.  This creates unnecessary GC.  Here is an example to demonstrate: https://go.dev/play/p/6fYrZTZiUeF

Might I suggest using a ring buffer instead.  [Here is one](https://github.com/gammazero/deque#deque) that can be pushed and popped and will reuse the same memory.